### PR TITLE
chore: add DocumentDelete api

### PR DESCRIPTION
**Pull request by arriqaaq**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -204,7 +204,7 @@ build/codegenv2:
 	  --go-grpc_out=require_unimplemented_servers=false,paths=source_relative:pkg/api/protomodel \
 	  --grpc-gateway_out=logtostderr=true,paths=source_relative:pkg/api/protomodel \
 	  --doc_out=pkg/api/protomodel --doc_opt=markdown,docs.md \
-	  --openapiv2_out=allow_merge=true,simple_operation_ids=true,disable_service_tags=true:pkg/api/openapi \
+	#   --openapiv2_out=allow_merge=true,simple_operation_ids=true,disable_service_tags=true:pkg/api/openapi \
```
Comment (jeroiraz) [inline Makefile:207]: spec needs to be regenerated, note the generation plugin has recently changed in base branch

File: embedded/document/engine.go [hunk 2]
```diff
@@ -938,3 +938,79 @@ func (e *Engine) readMetadataAndValue(key []byte, atTx uint64, skipIntegrityChec
...
+func (e *Engine) DeleteDocument(ctx context.Context, collectionName string, query *protomodel.Query) (err error) {
+	sqlTx, err := e.sqlEngine.NewTx(ctx, sql.DefaultTxOptions())
+	if err != nil {
+		return mayTranslateError(err)
+	}
+	defer sqlTx.Cancel()
+
+	table, err := getTableForCollection(sqlTx, collectionName)
+	if err != nil {
+		return mayTranslateError(err)
```
Comment (jeroiraz) [inline embedded/document/engine.go:951]: error translation should be responsability of the code directly using other packages, in this case it's a function defined in this package so it shouldn't be needed

File: embedded/document/engine.go [hunk 3]
```diff
@@ -938,3 +938,79 @@ func (e *Engine) readMetadataAndValue(key []byte, atTx uint64, skipIntegrityChec
...
+	}
+
+	idFieldName := docIDFieldName(table)
+
+	queryCondition, err := e.generateSQLExpression(ctx, query, table)
+	if err != nil {
+		return mayTranslateError(err)
```
Comment (jeroiraz) [inline embedded/document/engine.go:958]: same as above

File: embedded/document/engine.go [hunk 4]
```diff
@@ -938,3 +938,79 @@ func (e *Engine) readMetadataAndValue(key []byte, atTx uint64, skipIntegrityChec
...
+
+	queryStmt := sql.NewSelectStmt(
+		[]sql.Selector{sql.NewColSelector(collectionName, idFieldName)},
+		collectionName,
+		queryCondition,
+		nil,
+		nil,
+	)
+
+	r, err := e.sqlEngine.QueryPreparedStmt(ctx, sqlTx, queryStmt, nil)
```
Comment (jeroiraz) [inline embedded/document/engine.go:969]: it may be better to directly use sql delete statement

File: embedded/document/errors.go [hunk 5]
```diff
@@ -23,20 +23,21 @@ import (
...
+	ErrMultipleDocumentsFound   = errors.New("multiple documents found")
+	ErrDocumentNotFound         = errors.New("document not found")
+	ErrDocumentIDMismatch       = errors.New("document id mismatch")
+	ErrNoMoreDocuments          = errors.New("no more documents")
+	ErrFieldAlreadyExists       = errors.New("field already exists")
+	ErrFieldDoesNotExist        = errors.New("field does not exist")
+	ErrReservedFieldName        = errors.New("reserved field name")
+	ErrLimitedIndexCreation     = errors.New("index creation is only supported on empty collections")
+	ErrConflict                 = errors.New("conflict due to uniqueness contraint violation or read document was updated by another transaction")
+	ErrMultipleDocumentsMatched = errors.New("more than one document matched the query")
```
Comment (jeroiraz) [inline embedded/document/errors.go:40]: I'd think it's not required

File: embedded/sql/sql_tx.go [hunk 6]
```diff
@@ -142,6 +142,6 @@ func (sqlTx *SQLTx) Closed() bool {
 	return sqlTx.closed
 }
 
-func (sqlTx *SQLTx) delete(key []byte) error {
+func (sqlTx *SQLTx) Delete(key []byte) error {
```
Comment (jeroiraz) [inline embedded/sql/sql_tx.go:145]: I wouldn't expose this method, sql statement should not expose the low level key

File: pkg/api/proto/documents.proto [hunk 7]
```diff
@@ -226,6 +226,13 @@ message DocumentProofResponse {
   schema.VerifiableTxV2 verifiableTx = 5;
 }
 
+message DocumentDeleteRequest {
+  string collection = 1;
+  Query query = 2;
```
Comment (jeroiraz) [inline pkg/api/proto/documents.proto:231]: maybe we can add a `limit` , even an `offset`,  the sql delete statement already support it

File: pkg/server/documents_operations.go [hunk 8]
```diff
@@ -247,15 +246,15 @@ func (s *ImmuServer) DocumentSearch(ctx context.Context, req *protomodel.Documen
 	}
 
 	// read the next page of data from the paginated reader
-	results, err := resultReader.Read(ctx, int(req.PerPage))
-	if err != nil && err != sql.ErrNoMoreRows {
+	results, err := resultReader.ReadN(ctx, int(req.PerPage))
+	if err != nil && err != document.ErrNoMoreDocuments {
```
Comment (jeroiraz) [inline pkg/server/documents_operations.go:250]: errors.Is

File: pkg/server/documents_operations.go [hunk 9]
```diff
@@ -247,15 +246,15 @@ func (s *ImmuServer) DocumentSearch(ctx context.Context, req *protomodel.Documen
...
 		return nil, err
 	}
 
 	// update the pagination parameters for this query in the session
 	sess.UpdatePaginatedDocumentReader(searchID, req.Page, req.PerPage, int(pgreader.TotalRead)+len(results))
 
-	if err == sql.ErrNoMoreRows {
+	if err == document.ErrNoMoreDocuments {
```
Comment (jeroiraz) [inline pkg/server/documents_operations.go:257]: errors.Is

File: pkg/server/documents_operations.go [hunk 10]
```diff
@@ -276,3 +275,15 @@ func getSearchIDFromRequest(req *protomodel.DocumentSearchRequest) string {
 
 	return xid.New().String()
 }
+
+func (s *ImmuServer) DocumentDelete(ctx context.Context, req *protomodel.DocumentDeleteRequest) (*protomodel.DocumentDeleteResponse, error) {
+	db, err := s.getDBFromCtx(ctx, "DocumentDelete")
```
Comment (jeroiraz) [inline pkg/server/documents_operations.go:280]: may need to be added into permissions

File: Makefile (hunk 1 above)
Comment (arriqaaq) [reply Makefile:207]: fails for the swagger plugin

File: embedded/document/engine.go (hunk 2 above)
Comment (arriqaaq) [reply embedded/document/engine.go:951]: resolved

File: embedded/document/engine.go (hunk 4 above)
Comment (arriqaaq) [reply embedded/document/engine.go:969]: resolved

File: embedded/sql/sql_tx.go (hunk 6 above)
Comment (arriqaaq) [reply embedded/sql/sql_tx.go:145]: resolved

File: pkg/api/proto/documents.proto (hunk 7 above)
Comment (arriqaaq) [reply pkg/api/proto/documents.proto:231]: It's confusing, and imo, better to keep it simple for now.

Comment (arriqaaq) [reply pkg/api/proto/documents.proto:231]: This is single document delete, I am assuming we are in sync with that

**Pull request merged**